cdc/change_visitor.hh

```diff
@@ -0,0 +1,289 @@
+            return;
+        }
+    }
+
+    for (auto& cr: p.clustered_rows()) {
+        auto& r = cr.row();
+
+        if (r.marker().is_live() || !r.cells().empty()) {
+            v.clustered_row_cells(cr.key(), [&s, &r] (auto& crv) {
+                if (crv.finished()) {
+                    return;
+                }
+
+                auto& rm = r.marker();
+                if (rm.is_live()) {
+                    crv.marker(rm);
+
+                    if (crv.finished()) {
+                        return;
+                    }
+                }
+
+                inspect_row_cells<column_kind::regular_column>(s, r.cells(), crv);
+            });
+
+            if (v.finished()) {
+                return;
+            }
+        }
+
+        if (r.deleted_at()) {
+            auto t = r.deleted_at().tomb();
+            assert(t.timestamp != api::missing_timestamp);
+            v.clustered_row_delete(cr.key(), t);
+            if (v.finished()) {
+                return;
+            }
+        }
+    }
+
+    for (auto& rt: p.row_tombstones()) {
+        assert(rt.tomb.timestamp != api::missing_timestamp);
+        v.range_delete(rt);
+        if (v.finished()) {
+            return;
+        }
+    }
+
+    if (p.partition_tombstone()) {
+        v.partition_delete(p.partition_tombstone());
+    }
+}
+
+template <column_kind ckind, RowCellsVisitor V>
+void inspect_row_cells(const schema& s, const row& r, V& v) {
+    r.for_each_cell_until([&s, &v] (column_id id, const atomic_cell_or_collection& acoc) {
+        auto& cdef = s.column_at(ckind, id);
+
+        if (cdef.is_atomic()) {
+            auto cell = acoc.as_atomic_cell(cdef);
+            if (cell.is_live()) {
+                v.live_atomic_cell(cdef, cell);
+            } else {
+                v.dead_atomic_cell(cdef, cell);
+            }
+
+            if (v.finished()) {
+                return stop_iteration::yes;
+            }
+            return stop_iteration::no;
+        }
+
+        auto stop = stop_iteration::no;
+        acoc.as_collection_mutation().with_deserialized(*cdef.type, [&v, &cdef, &stop] (collection_mutation_view_description view) {
+            v.collection_column(cdef, [&view] (auto& ccv) {
+                if (ccv.finished()) {
+                    return;
+                }
+
+                if (view.tomb) {
+                    ccv.collection_tombstone(view.tomb);
+                    if (ccv.finished()) {
+                        return;
+                    }
+                }
+
+                for (auto& [key, cell]: view.cells) {
+                    if (cell.is_live()) {
+                        ccv.live_collection_cell(key, cell);
+                    } else {
+                        ccv.dead_collection_cell(key, cell);
+                    }
+
+                    if (ccv.finished()) {
+                        return;
+                    }
+                }
+            });
+
+            if (v.finished()) {
```

Comment:
Maybe just stop = stop_iteration(v.finished())? That would totally remove the branch.